### PR TITLE
feat: more Results, fewer panics, always have backtraces

### LINKED ISSUE #564: Enforce no unwrap()
Enable https://rust-lang.github.io/rust-clippy/master/index.html#/unwrap_used and fix all the cases

### PR DESCRIPTION
**Pull request by lwwmanning**
fixes #564 

The goals of this PR is to ensure the following:
- reasonable developer ergonomics / not too verbose
- return a `VortexResult` with a well-formed error from most fallible functions
- panics are allowed in cases of programmer error / invariants being violated (i.e., panics should correspond logically to places where we might write an `assert`)
- ensure a good error message and a backtrace in case of panic

To accomplish this, this PR does several things. 

First, it adds much stricter clippy lints (e.g., deny on `unwrap`, `expect`, `panic`; forbids fallible `From` impls; etc.). 

Second, I did an audit to make functions that are truly fallible return `VortexResult` instead of panicking. The remaining ones that I saw but didn't fix are tracked under #765 

Third, for places where we truly want to panic (specifically because an internal invariant has been violated, typically indicating programmer error), this PR adds:
- a new macro (`vortex_panic!`)
- `VortexUnwrap` trait, which is implemented only on `VortexResult` and adds a `vortex_unwrap()` function
- `VortexExpect` trait, which is implemented on `VortexResult` and `Option<T>` and adds a `vortex_expect` function that takes a string literal

Taken together, the main downsides are that we have a special snowflake unwrap/expect instead of the traditional ones, and we don't have an effective lint rule to prevent calling those in functions that return Result... but I think the latter at least is acceptable (I know @AdamGS finds those lints overly restrictive anyway).

ALTERNATIVELY, a middle ground would be to replace calls to `vortex_unwrap` and `vortex_expect` with `expect` and un-deny `expect_used`. This would be more "typical" Rust but with the disadvantage that we wouldn't get backtraces by default in the cases where we have bugs. Or we could do `unwrap_or_else(|err| vortex_panic!(err))` or `ok_or_else(|| vortex_panic!(...))` everywhere, but that's pretty verbose.

### CODE REVIEW

**Comment by lwwmanning** [comment]
FLUP tracked in #765

File: Cargo.toml [hunk 1]
```diff
@@ -171,11 +171,31 @@ unsafe_op_in_unsafe_fn = "deny"
 
 [workspace.lints.clippy]
 all = { level = "deny", priority = -1 }
+as_ptr_cast_mut = { level = "deny" }
```
Comment (robert3005) [inline Cargo.toml:174]: Is there a collection we can list here instead of enumerating all of them? Why isn't this `all`?

Comment (robert3005) [reply Cargo.toml:174]: `all` means `all lints that are on by default (correctness, suspicious, style, complexity, perf)`

Comment (robert3005) [reply Cargo.toml:174]: but seems like you could just deny other groups instead of individual lints

Comment (lwwmanning) [reply Cargo.toml:174]: this is a pick-and-choose from `pedantic` and `restriction`; those two groups are both very noisy, so the clippy docs don't recommend turning them on by default

**Comment by gatesn** [comment]
What do you mean we wouldn’t get backtraces by default using builtin expect?

File: bench-vortex/src/lib.rs [hunk 2]
```diff
@@ -190,7 +190,8 @@ pub fn compress_taxi_data() -> Array {
     let chunks = reader
         .into_iter()
         .map(|batch_result| batch_result.unwrap())
-        .map(Array::from)
+        .map(Array::try_from)
+        .map(Result::unwrap)
```
Comment (robert3005) [inline bench-vortex/src/lib.rs:194]: should collect errors and unwrap at the ends

File: bench-vortex/src/data_downloads.rs [hunk 3]
```diff
@@ -45,7 +45,7 @@ pub fn data_vortex_uncompressed(fname_out: &str, downloaded_data: PathBuf) -> Pa
         let array = ChunkedArray::try_new(
             reader
                 .into_iter()
-                .map(|batch_result| Array::from(batch_result.unwrap()))
+                .map(|batch_result| Array::try_from(batch_result.unwrap()).unwrap())
```
Comment (robert3005) [inline bench-vortex/src/data_downloads.rs:48]: collect errors and unwrap at the end

File: bench-vortex/src/reader.rs [hunk 4]
```diff
@@ -101,7 +102,7 @@ pub fn compress_parquet_to_vortex(parquet_path: &Path) -> VortexResult<ChunkedAr
     let chunks = reader
         .map(|batch_result| batch_result.unwrap())
         .map(|record_batch| {
-            let vortex_array = Array::from(record_batch);
+            let vortex_array = Array::try_from(record_batch).unwrap();
```
Comment (robert3005) [inline bench-vortex/src/reader.rs:105]: should collect errors and unwrap at the end

File: bench-vortex/src/tpch/mod.rs [hunk 5]
```diff
@@ -247,8 +247,8 @@ async fn register_vortex_file(
         let sts = record_batches
             .iter()
             .cloned()
-            .map(Array::from)
-            .map(|a| a.into_struct().unwrap())
+            .map(Array::try_from)
+            .map(|a| a.unwrap().into_struct().unwrap())
```
Comment (robert3005) [inline bench-vortex/src/tpch/mod.rs:251]: should collect errors

**Review by AdamGS — APPROVED** [review]
LGTM, seems to me like most rust projects end up having something like this and I think it solves most things that bothered me with the current clippy config while still maintaining error context and backtraces

**Review by robert3005 — APPROVED** [review]
I left some comments about lazy unwraps. I am not convinced this is the best way of doing things but don't know how to do anything better

**Comment by lwwmanning** [comment]
> What do you mean we wouldn’t get backtraces by default using builtin expect?

We're a library. Unless the actual binary sets `RUST_BACKTRACE=1`, regular expect would just show us something like:

```
thread 'main' panicked at 'here's the message I gave to expect', foo.rs:100:10
note: Run with `RUST_BACKTRACE=1` for a backtrace.

```

Which is just not that helpful

**Pull request merged**